lib/upload.js

```diff
@@ -923,15 +924,24 @@ function shouldRetry(err, retryAttempt, options) {
   // - retryDelays option is set
   // - we didn't exceed the maxium number of retries, yet, and
   // - this error was caused by a request or it's response and
-  // - the error is server error (i.e. no a status 4xx or a 409 or 423) and
+  // - the error is server error (i.e. no a status 4xx or a 409 or 423) or
+  // a onShouldRetry is specified and returns true
   // - the browser does not indicate that we are offline
   let status = err.originalResponse ? err.originalResponse.getStatus() : 0;
-  let isServerError = !inStatusCategory(status, 400) || status === 409 || status === 423;
+
+  // define as function so we only call it if really needed
+  let doRetry = () => {
```

Comment:
Can you please refactor the code to not use an anonymous function here? I have something like this in mind:

```js
if (retryDelays == null || retryAttempt < retryDelays.lenght || originalRequest == null) {
  return false;
}

if (options && typeof options.onShouldRetry === "function") {
  return options.onShouldRetry(err, retryAttempt, options);
}

return !inStatusCategory(status, 400) || status === 409 || status === 423 && online();
```

That's easier to read IMO.